lib/join/joinclient/oracle/oracle.go

```diff
@@ -0,0 +1,156 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// Package oracle includes methods required for clients to join a Teleport
+// cluster using the Oracle join method.
+package oracle
+
+import (
+	"context"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"io"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/oracle/oci-go-sdk/v65/common"
+	"github.com/oracle/oci-go-sdk/v65/common/auth"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/utils/keys"
+	"github.com/gravitational/teleport/lib/join/internal/messages"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// SolveChallenge solves an Oracle join challenge.
+func SolveChallenge(ctx context.Context, httpClient utils.HTTPDoClient, challenge *messages.OracleChallenge) (*messages.OracleChallengeSolution, error) {
+	signedRootCAReq, err := makeSignedRootCAReq(httpClient)
+	if err != nil {
+		return nil, trace.Wrap(err, "making signed root CA request")
+	}
+
+	cert, err := fetchIMDS(ctx, httpClient, "cert.pem")
+	if err != nil {
+		return nil, trace.Wrap(err, "fetching instance cert from IMDS")
+	}
+
+	intermediate, err := fetchIMDS(ctx, httpClient, "intermediate.pem")
+	if err != nil {
+		return nil, trace.Wrap(err, "fetching intermediate CA certs from IMDS")
+	}
+
+	keyPEM, err := fetchIMDS(ctx, httpClient, "key.pem")
+	if err != nil {
+		return nil, trace.Wrap(err, "fetching instance private key from IMDS")
+	}
+
+	key, err := keys.ParsePrivateKey(keyPEM)
+	if err != nil {
+		return nil, trace.Wrap(err, "parsing instance private key")
+	}
+
+	signature, err := SignChallenge(key, challenge.Challenge)
+	if err != nil {
+		return nil, trace.Wrap(err, "signing challenge")
+	}
+
+	return &messages.OracleChallengeSolution{
+		Cert:            cert,
+		Intermediate:    intermediate,
+		Signature:       signature,
+		SignedRootCAReq: signedRootCAReq,
+	}, nil
+}
+
+// SignChallenge signs a challenge with a given instance private key.
+func SignChallenge(key crypto.Signer, challenge string) ([]byte, error) {
+	if _, ok := key.Public().(*rsa.PublicKey); !ok {
+		return nil, trace.BadParameter("only RSA keys are supported")
+
+	}
+	signerOpts := &rsa.PSSOptions{SaltLength: rsa.PSSSaltLengthEqualsHash, Hash: crypto.SHA256}
+	signature, err := crypto.SignMessage(key, rand.Reader, []byte(challenge), signerOpts)
+	return signature, trace.Wrap(err)
+}
+
+func makeSignedRootCAReq(client utils.HTTPDoClient) ([]byte, error) {
```

Comment:
I know this matches the RFD, but I wonder if we can/should make the join method more flexible by moving the request construction logic over to the Teleport server. The client would simply sign the request received from upstream server, assuming it deems the request acceptable.